include/engine/engine.hpp

```diff
@@ -30,38 +32,162 @@ namespace osrm
 namespace engine
 {
 
-class Engine final
+class EngineInterface
+{
+  public:
+    virtual ~EngineInterface(){};
+    virtual Status Route(const api::RouteParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Table(const api::TableParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Nearest(const api::NearestParameters &parameters,
+                           util::json::Object &result) const = 0;
+    virtual Status Trip(const api::TripParameters &parameters,
+                        util::json::Object &result) const = 0;
+    virtual Status Match(const api::MatchParameters &parameters,
+                         util::json::Object &result) const = 0;
+    virtual Status Tile(const api::TileParameters &parameters, std::string &result) const = 0;
+};
+
+template <typename AlgorithmT> class Engine final : public EngineInterface
 {
   public:
-    explicit Engine(const EngineConfig &config);
+    explicit Engine(const EngineConfig &config)
+        : route_plugin(config.max_locations_viaroute),       //
+          table_plugin(config.max_locations_distance_table), //
+          nearest_plugin(config.max_results_nearest),        //
+          trip_plugin(config.max_locations_trip),            //
+          match_plugin(config.max_locations_map_matching),   //
+          tile_plugin()                                      //
+
+    {
+        if (config.use_shared_memory)
+        {
+            facade_provider = std::make_unique<WatchingProvider<AlgorithmT>>();
+        }
+        else
+        {
+            facade_provider =
+                std::make_unique<ImmutableProvider<AlgorithmT>>(config.storage_config);
+        }
+    }
 
     Engine(Engine &&) noexcept = delete;
     Engine &operator=(Engine &&) noexcept = delete;
 
     Engine(const Engine &) = delete;
     Engine &operator=(const Engine &) = delete;
+    virtual ~Engine(){};
+
+    Status Route(const api::RouteParameters &params,
```

Comment:
maybe a function abstracting over all of this?

    template <typename Plugin, typename Algo>
    Status Run(FacadeProvider fp, Heaps heaps, Plugin plugin, Params params, Result& ret) {
      facade = fp.get();
      algo = RoutingAlgorithms<Algo>{heaps, facade};
      return plugin.HandleRequests(facade, algo, params, result);
    }

(`HandleRequest` is not virtual in base so we need the plugin template here)

and then for each Service function do

    return Run<Table, AlgorithmT>(facade_provider, heaps, table_plugin, params, result);

Hm we could be even fancier here, all we do is a type-level loop over Service and ServiceParams

    for_each({{Table, TableParams}, {Route, RouteParams}, [](){..})